dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java

```
@@ -239,46 +243,75 @@ public PerHostClientFactory perHost(String serviceName) {
+                        selfAddresses = Set.of();
+                    }
+
+                    ImmutableMap.Builder<PerHostTarget, Channel> map = ImmutableMap.builder();
                     for (int i = 0; i < targetUris.size(); i++) {
                         TargetUri targetUri = targetUris.get(i);
                         ServiceConfiguration singleUriServiceConf = ServiceConfiguration.builder()
                                 .from(serviceConfiguration)
                                 .uris(ImmutableList.of(targetUri.uri()))
                                 .build();
+                        boolean isSelf = targetUri
+                                .resolvedAddress()
+                                .map(selfAddresses::contains)
+                                .orElse(false);
 
                         // subtle gotcha here is that every single one of these has the same channelName,
                         // which means metrics like the QueuedChannel counter will end up being the sum of all of them.
-                        list.add(cache.getNonReloadingChannel(
-                                params,
-                                singleUriServiceConf,
-                                ImmutableList.of(targetUri),
-                                channelName,
-                                OptionalInt.of(i)));
+                        map.put(
+                                new PerHostTarget(targetUri, isSelf),
+                                cache.getNonReloadingChannel(
+                                        params,
+                                        singleUriServiceConf,
+                                        ImmutableList.of(targetUri),
+                                        channelName,
+                                        OptionalInt.of(i)));
                     }
-                    return list.build();
+                    return map.buildKeepingLast();
```

I think we want this to fail loudly/catastrophically if something goes wrong

---

Is this safe? There's no guarantee that the target URIs are unique.

---

Hmm, I suppose an example might look something like this:
```yaml
uris:
  - https://target/api
  - https://target/api
```

In such a case, I think I'd prefer if we pruned out duplicates earlier on, otherwise we risk effectively doubling the concurrency limits on repeated values. I think we can improve things across the board by updating `ReloadingClientFactory.getTargetUris` with something like this to retain order, but filter duplicates:
```diff
-return ImmutableList.copyOf(targetUris);
+return ImmutableSet.copyOf(targetUris).toList();
```